sdk/program/src/message/sanitized.rs

```diff
@@ -0,0 +1,599 @@
+use {
+    crate::{
+        fee_calculator::FeeCalculator,
+        hash::Hash,
+        instruction::{CompiledInstruction, Instruction},
+        message::{MappedAddresses, MappedMessage, Message, MessageHeader},
+        pubkey::Pubkey,
+        sanitize::{Sanitize, SanitizeError},
+        secp256k1_program,
+        serialize_utils::{append_slice, append_u16, append_u8},
+    },
+    bitflags::bitflags,
+    std::convert::TryFrom,
+    thiserror::Error,
+};
+
+/// Sanitized message of a transaction which includes a set of atomic
+/// instructions to be executed on-chain
+#[derive(Debug, Clone)]
+pub enum SanitizedMessage {
+    /// Sanitized legacy message
+    Legacy(Message),
+    /// Sanitized version #0 message with mapped addresses
+    V0(MappedMessage),
+}
+
+#[derive(PartialEq, Debug, Error, Eq, Clone)]
+pub enum SanitizeMessageError {
+    #[error("index out of bounds")]
+    IndexOutOfBounds,
+    #[error("value out of bounds")]
+    ValueOutOfBounds,
+    #[error("invalid value")]
+    InvalidValue,
+    #[error("duplicate account key")]
+    DuplicateAccountKey,
+}
+
+impl From<SanitizeError> for SanitizeMessageError {
+    fn from(err: SanitizeError) -> Self {
+        match err {
+            SanitizeError::IndexOutOfBounds => Self::IndexOutOfBounds,
+            SanitizeError::ValueOutOfBounds => Self::ValueOutOfBounds,
+            SanitizeError::InvalidValue => Self::InvalidValue,
+        }
+    }
+}
+
+impl TryFrom<Message> for SanitizedMessage {
+    type Error = SanitizeMessageError;
+    fn try_from(message: Message) -> Result<Self, Self::Error> {
+        message.sanitize()?;
+
+        let sanitized_msg = Self::Legacy(message);
+        if sanitized_msg.has_duplicates() {
+            return Err(SanitizeMessageError::DuplicateAccountKey);
+        }
```

Comment:
fyi @t-nelson I just added this forgotten dupe check which is currently inconsequential but could have caused issues later